src/libmcrx.c

```diff
@@ -1147,144 +1170,146 @@ MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_apply(struct mcrx_ctx *ctx,
+        && global_address->addr_type == MCRX_ADDR_TYPE_V6
+        && memcmp(&entry->global_addrs.v6.source,
+            &global_address->addrs.v6.source, sizeof(struct in6_addr)) == 0
+        && memcmp(&entry->global_addrs.v6.group,
+            &global_address->addrs.v6.group, sizeof(struct in6_addr)) == 0) {
       return entry;
     }
   }
   return entry;
 }
 
 /**
- * mcrx_mnat_ctx_add_entry:
+ * mcrx_mnat_ctx_add_or_update_mapping:
  *
  * Note: unresolved local source and local group needed to be populated as 0.0.0.0
  * Add the mcrx mnat entry.
  *
  **/
-MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_add_entry(
-    struct mcrx_mnat_ctx *mnatctx, const char *global_source,
-    const char *global_group, const char *local_source, const char *local_group) {
-  if (global_source == NULL || global_group == NULL || local_source == NULL
-      || local_group == NULL) {
+enum mcrx_error_code mcrx_mnat_ctx_add_or_update_mapping(
+    struct mcrx_mnatmap *mnatmap, const struct mcrx_subscription_config *global_address,
+    const struct mcrx_subscription_config *local_address) {
+  bool add_entry = false;
+  if (global_address == NULL || local_address == NULL) {
     return MCRX_ERR_NULLARG;
   }
-
-  struct mcrx_mnat_entry *entry = mcrx_mnat_ctx_find_entry(mnatctx,
-      global_source, global_group);
-  if (entry != NULL) {
-    return MCRX_ERR_ALREADY_EXISTED;
-  }
-  entry = calloc(1, sizeof(struct mcrx_mnat_entry));
-  if (!entry) {
-    return MCRX_ERR_NOMEM;
-  }
-  int ret;
-  ret = inet_pton(AF_INET, global_source, &entry->global_addrs.v4.source);
-  if (ret > 0) {
-    ret = inet_pton(AF_INET, global_group, &entry->global_addrs.v4.group);
-    if (ret > 0) {
-      ret = inet_pton(AF_INET, local_source, &entry->local_addrs.v4.source);
-      if (ret > 0) {
-        ret = inet_pton(AF_INET, local_group, &entry->local_addrs.v4.group);
-        if (ret > 0) {
-          entry->addr_type = MCRX_ADDR_TYPE_V4;
-        }
-      }
+  struct mcrx_mnat_entry *entry = mcrx_mnat_ctx_find_entry(mnatmap,
+      global_address);
+  if (entry == NULL) {
+    entry = calloc(1, sizeof(struct mcrx_mnat_entry));
+    if (!entry) {
+      return MCRX_ERR_NOMEM;
     }
-  }
-
-  ret = inet_pton(AF_INET, global_source, &entry->global_addrs.v6.source);
-  if (ret > 0) {
-    ret = inet_pton(AF_INET, global_group, &entry->global_addrs.v6.group);
-    if (ret > 0) {
-      ret = inet_pton(AF_INET, local_source, &entry->local_addrs.v6.source);
-      if (ret > 0) {
-        ret = inet_pton(AF_INET, local_group, &entry->local_addrs.v6.group);
-        if (ret > 0) {
-          entry->addr_type = MCRX_ADDR_TYPE_V6;
-        }
-      }
+    add_entry = true;
+  }
+  if (global_address->addr_type == MCRX_ADDR_TYPE_V4 && local_address->addr_type == MCRX_ADDR_TYPE_V4) {
+      entry->addr_type = global_address->addr_type;
+      memcpy(&entry->global_addrs.v4.source, &global_address->addrs.v4.source, sizeof(struct in_addr));
```

Comment:
if we do the memcpy of the global_addrs inside the calloc, we can assume the global_addrs is already set (because the find only happens if they're equal), so we don't have to do the memcpy here for the case there was already a mapping (also in the v6 case).

Reply:
@GrumpyOldTroll mcrx_mnat_ctx_add_or_update_mapping
calloc only allocated and sets allocated memory to zero.

Reply:
Sorry, I didn't mean "inside the calloc", I meant immediately after the calloc and the null check that follows it, inside the block that contains the calloc.

My point is that whenever the entry was created, the global address and group are known at that time and should be set immediately.  Whenever the entry was not created, the global address and group were known during the lookup and exactly match the input parameters.  Therefore, the memcpy into the entry's global addresses should be done after the object is allocated, rather than after the object has been looked up.

Otherwise here we're doing a pointless memcpy of the global addresses whenever the lookup succeeded.